<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Using Jupyter (Good), Exercises</span></div>

**Exercise 1 (Learning to interrupt)**

In [1]:
while True:
    print('*', end='')



****************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************

KeyboardInterrupt: 

**Exercise 2 (Resetting and Restarting the Kernal)**

In [2]:
print = 'PRINT'

In [3]:
print('Hello World')

TypeError: 'str' object is not callable

In [2]:
print('Hello World')

## Yay!

Hello World


**Exercise 3 (Timing Stuff)**

In [1]:
%%timeit -n 5000
squares = [i**2 for i in range (10)]

print(squares)
##My cell magic is not working, i tried line magic also but cannnot
##Ans: you need to specify the number of loops it can run so it wont run until the insane number of default loops

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[

In [ ]:
%%timeit -n 5000

squares = list()
for i in range(10):
    squares.append(i**2)

859 ns ± 158 ns per loop (mean ± std. dev. of 7 runs, 5,000 loops each)


The slowest run took 40.79 times longer than the fastest. This could mean that an intermediate result is being cached.
153 μs ± 207 μs per loop (mean ± std. dev. of 7 runs, 5,000 loops each)


859 ns ± 158 ns per loop (mean ± std. dev. of 7 runs, 5,000 loops each)


| Description  |  Time per loop   |    Meaning |
| :--- | :--: | ---: |
| Immediately executing command without defining it as a list first  |  186 μs ± 257 μs per loop   |   It is faster to run a line of code when you define it as a list first instead of making it look shorter and using square brackets only |
| Defining it as a list first  |  859 ns ± 158 ns   |   It is ~216x faster to define it as a list first


## Footnotes